src/error.rs

```diff
@@ -0,0 +1,239 @@
+        source: SourceFile<'s>,
+    ) -> Result<T, CompilerError<'s, E>>;
+}
+
+impl<T, E> ResultExt<T, E> for Result<T, E>
+where
+    E: Into<CompilerErrorKind>,
+{
+    fn map_err_fatal<'s>(
+        self,
+        ectx: &mut ErrorContext<'s>,
+        source: SourceFile<'s>,
+    ) -> Result<T, CompilerError<'s, E>> {
+        match self {
+            Ok(i) => Ok(i),
+            Err(e) => ectx.fatal(e, source)?,
+        }
+    }
+}
+
+/// The toplevel compiler error enum. This is what all errors finally turn in to, usually
+/// with transparent wrappers to an actual Diagnostic.
+#[derive(Diagnostic, Error, Debug, Clone, PartialEq)]
+pub enum CompilerErrorKind {
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Lex(#[from] LexError),
+
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    Parse(#[from] ParseError),
+}
+
+impl CompilerErrorKind {
+    pub fn to_miette(&self, source: &SourceFile) -> miette::Report {
+        miette::Report::new(self.clone()).with_source_code(source.named_source())
+    }
+}
+
+/// The error type returned out of most functions in the compiler.
+/// Contains a number of "recoverable errors" in the error context, and then one last which was fatal.
+#[derive(PartialEq, Clone)]
+pub struct CompilerError<'s, E: Into<CompilerErrorKind>> {
+    related: Vec<(CompilerErrorKind, SourceFile<'s>)>,
+    fatal: (E, SourceFile<'s>),
+}
+
+impl<'s, E> CompilerError<'s, E>
+where
+    E: Into<CompilerErrorKind>,
+{
+    /// Get the type of fatal error inside
+    #[allow(unused)] // used in tests
+    pub fn get_fatal(&self) -> &E {
+        &self.fatal.0
+    }
+
+    pub fn find_related(
+        &self,
+        f: impl Fn(&CompilerErrorKind) -> bool,
+    ) -> Option<&CompilerErrorKind> {
+        self.related.iter().map(|i| &i.0).find(|x| f(x))
+    }
+
+    #[allow(unused)] // used in tests
+    pub fn contains(&self, e: impl Into<CompilerErrorKind>) -> bool
+    where
+        E: Clone,
+    {
+        let e = e.into();
+        self.find_related(|x| x == &e).is_some() || self.fatal.0.clone().into() == e
+    }
+
+    /// Turns a generic error of any `E` into an error of [CompilerErrorKind](CompilerErrorKind)
+    pub fn make_generic(self) -> CompilerError<'s, CompilerErrorKind> {
+        CompilerError {
+            related: self.related,
+            fatal: (self.fatal.0.into(), self.fatal.1),
+        }
+    }
+}
+
+impl<'s, E> Debug for CompilerError<'s, E>
+where
+    E: Into<CompilerErrorKind> + Diagnostic + Clone,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        for (e, source) in &self.related {
+            writeln!(f, "{:?}", e.clone().to_miette(source))?;
+        }
+
+        let (e, source) = &self.fatal;
+        write!(f, "{:?}", e.clone().into().to_miette(source))
+    }
+}
+
+/// Holds the current compiler state of recoverable errors until a fatal error is thrown.
+pub struct ErrorContext<'s> {
+    errors: Vec<(CompilerErrorKind, SourceFile<'s>)>,
```

Comment:
We can group compiler errors by their source files.  In fact, until the name resolution / type-checking stage, all processing is happening per-file, so I think it's not a bad idea to have a concept of "current source file" in this type (or in a wrapper around this type).